File: packages/dual_cameras_ios/ios/Classes/DualCamerasPlugin.swift
import AVFoundation
import Flutter

/// iOS implementation of dual_cameras: AVCaptureMultiCamSession + Metal
/// compositor + AVAssetWriter, unified for preview / video / photo
/// (ARCHITECTURE.md §5). Cannot be compiled on Linux CI; verified by review and
/// on-device.
public final class DualCamerasPlugin: NSObject, FlutterPlugin, DualCameraHostApi {

    private let textureRegistry: FlutterTextureRegistry
    private let flutterApi: DualCameraFlutterApi

    private var previewTexture: DualCameraTexture?
    private var textureId: Int64?
    private var compositor: MetalCompositor?
    private var session: DualCameraSession?
    private var layout = CompositeLayout(
        pictureInPicture: true, primaryFront: false,
        insetRect: CGRect(x: 0.7, y: 0.68, width: 0.28, height: 0.28),
        cornerRadiusPx: 18, mirrorFront: true, splitVertical: false)
    private var hevc = false

    init(textureRegistry: FlutterTextureRegistry, flutterApi: DualCameraFlutterApi) {
        self.textureRegistry = textureRegistry
        self.flutterApi = flutterApi
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        let instance = DualCamerasPlugin(
            textureRegistry: registrar.textures(),
            flutterApi: DualCameraFlutterApi(binaryMessenger: messenger))
        DualCameraHostApiSetup.setUp(binaryMessenger: messenger, api: instance)
        // Debug-tuning channel (IOS_HANDOFF §6) — mirrors the Android
        // `dual_cameras/debug` channel; the shared example panel drives
        // it. Gate behind kDebugMode / strip before pub.dev publish.
        let debug = FlutterMethodChannel(
            name: "dual_cameras/debug", binaryMessenger: messenger)
        debug.setMethodCallHandler { [weak instance] call, result in
            instance?.handleDebug(call, result) ?? result(FlutterMethodNotImplemented)
        }
        registrar.publish(instance)
    }

    private func handleDebug(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        guard let session = session else {
            result(FlutterError(code: "not_init", message: "initialize() first", details: nil))
            return
        }
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "setRotationOffset":
            let front = args["front"] as? Bool ?? true
            let offset = args["offset"] as? Int ?? 0
            session.setRotationOffset(isFront: front, degrees: offset)
            result(nil)
        case "setAspectOverride":
            let front = args["front"] as? Bool ?? true
            let aspect = Float(args["aspect"] as? Double ?? 0)
            session.setAspectOverride(isFront: front, aspect: aspect)
            result(nil)
        case "setMirrorFront":
            let on = args["on"] as? Bool ?? true
            layout.mirrorFront = on
            session.setLayout(layout)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - DualCameraHostApi

    func probeSupport(completion: @escaping (Result<CameraCapabilities, Error>) -> Void) {
        completion(.success(detectCapabilities()))
    }

    func initialize(config: RecordingConfig,
                    completion: @escaping (Result<InitResult, Error>) -> Void) {
        let caps = detectCapabilities()
        let (w, h) = resolution(config.resolution)
        hevc = config.codec == .hevc
        layout = toLayout(config.layout, width: w, height: h)

        guard let compositor = MetalCompositor(width: w, height: h) else {
            completion(.failure(NSError(domain: "dual_cameras", code: 2,
                userInfo: [NSLocalizedDescriptionKey: "Metal init failed"])))
            return
        }
        self.compositor = compositor

        let texture = DualCameraTexture()
        let id = textureRegistry.register(texture)
        previewTexture = texture
        textureId = id

        let session = DualCameraSession(
            compositor: compositor, layout: layout, recordAudio: config.recordAudio)
        session.onFrame = { [weak self] buffer in
            guard let self = self, let id = self.textureId else { return }
            self.previewTexture?.push(buffer)
            self.textureRegistry.textureFrameAvailable(id)
        }
        session.onThermal = { [weak self] state in
            self?.flutterApi.onThermal(level: Self.mapThermal(state)) { _ in }
        }
        session.onError = { [weak self] message in
            self?.flutterApi.onError(code: "session", message: message) { _ in }
        }
        self.session = session
        session.configureAndStart()

        flutterApi.onReady(textureId: id) { _ in }
        completion(.success(InitResult(textureId: id, capabilities: caps)))
    }

    func startRecording(completion: @escaping (Result<Void, Error>) -> Void) {
        guard let session = session else { return completion(notReady()) }
        let url = tempURL("mp4")
        session.startRecording(url: url, hevc: hevc)
        flutterApi.onRecordingStarted { _ in }
        completion(.success(()))
    }

    func stopRecording(completion: @escaping (Result<String, Error>) -> Void) {
        guard let session = session else { return completion(notReady()) }
        session.stopRecording { [weak self] url in
            if let url = url {
                self?.flutterApi.onRecordingStopped(path: url.path) { _ in }
                completion(.success(url.path))
            } else {
                completion(.failure(NSError(domain: "dual_cameras", code: 3)))
            }
        }
    }

    func takePhoto(completion: @escaping (Result<String, Error>) -> Void) {
        guard let session = session else { return completion(notReady()) }
        let url = tempURL("jpg")
        session.takePhoto(url: url) { ok in
            ok ? completion(.success(url.path))
               : completion(.failure(NSError(domain: "dual_cameras", code: 4)))
        }
    }

    func swapPrimary(completion: @escaping (Result<Void, Error>) -> Void) {
        layout.primaryFront.toggle()
        session?.setLayout(layout)
        completion(.success(()))
    }

    func setLayout(layout: LayoutConfig, completion: @escaping (Result<Void, Error>) -> Void) {
        let w = compositor?.width ?? 1280
        let h = compositor?.height ?? 720
        self.layout = toLayout(layout, width: w, height: h)
        session?.setLayout(self.layout)
        completion(.success(()))
    }

    func dispose(completion: @escaping (Result<Void, Error>) -> Void) {
        session?.stop()
        session = nil
        compositor = nil
        if let id = textureId { textureRegistry.unregisterTexture(id) }
        previewTexture = nil
        textureId = nil
        completion(.success(()))
    }

    // MARK: - helpers

    private func notReady<T>() -> Result<T, Error> {
        .failure(NSError(domain: "dual_cameras", code: 1,
            userInfo: [NSLocalizedDescriptionKey: "not initialized"]))
    }

    private func tempURL(_ ext: String) -> URL {
        URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("dcr_\(Int(Date().timeIntervalSince1970 * 1000)).\(ext)")
    }

    /// Portrait canvas (vertical video): width < height. The cameras deliver
    /// landscape buffers; the Metal compositor rotates them upright into this
    /// portrait canvas (matches Android's `resolutionFor`).
    private func resolution(_ r: DualResolution) -> (Int, Int) {
        switch r {
        case .sd480: return (480, 854)
        case .hd720: return (720, 1280)
        case .hd1080: return (1080, 1920)
        }
    }

    private func toLayout(_ l: LayoutConfig, width: Int, height: Int) -> CompositeLayout {
        let scale = CGFloat(l.insetScale)
        let mx = CGFloat(l.margin) / CGFloat(width)
        let my = CGFloat(l.margin) / CGFloat(height)
        var rect: CGRect
        switch l.insetCorner {
        case .topLeft: rect = CGRect(x: mx, y: my, width: scale, height: scale)
        case .topRight: rect = CGRect(x: 1 - mx - scale, y: my, width: scale, height: scale)
        case .bottomLeft: rect = CGRect(x: mx, y: 1 - my - scale, width: scale, height: scale)
        case .bottomRight: rect = CGRect(x: 1 - mx - scale, y: 1 - my - scale, width: scale, height: scale)
        }
        return CompositeLayout(
            pictureInPicture: l.mode == .pictureInPicture,
            primaryFront: l.primary == .front,
            insetRect: rect,
            cornerRadiusPx: Float(l.cornerRadius),
            mirrorFront: l.mirrorFront,
            splitVertical: l.mode == .splitVertical)
    }

    private static func mapThermal(_ state: ProcessInfo.ThermalState) -> ThermalLevel {
        switch state {
        case .nominal: return .nominal
        case .fair: return .fair
        case .serious: return .serious
        case .critical: return .critical
        @unknown default: return .nominal
        }
    }

    private func detectCapabilities() -> CameraCapabilities {
        if #available(iOS 13.0, *) {
            if AVCaptureMultiCamSession.isMultiCamSupported {
                return CameraCapabilities(isSupported: true, maxWidth: 1920, maxHeight: 1080)
            }
            return CameraCapabilities(isSupported: false, reason: .noConcurrentCamera)
        }
        return CameraCapabilities(isSupported: false, reason: .osTooOld)
    }
}

File: packages/dual_cameras_android/android/src/main/kotlin/com/romanslack/dual_cameras_android/DualCamerasPlugin.kt
package com.romanslack.dual_cameras_android

import android.content.Context
import android.content.pm.PackageManager
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.PowerManager
import android.view.Surface
import java.util.concurrent.Executor
import com.romanslack.dual_cameras_android.camera.CameraSource
import com.romanslack.dual_cameras_android.gl.CompositeLayout
import com.romanslack.dual_cameras_android.pipeline.RenderThread
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry
import java.io.File

/**
 * Android implementation of dual_cameras.
 *
 * Capability detection + Pigeon wiring (Phase 0), and the unified GLES
 * compositor -> MediaCodec/MediaMuxer recording pipeline (Phase 1). Photo
 * capture and the perf HUD land in Phase 2; see ARCHITECTURE.md.
 */
class DualCamerasPlugin : FlutterPlugin, DualCameraHostApi {

  private var context: Context? = null
  private var textureRegistry: TextureRegistry? = null
  private var flutterApi: DualCameraFlutterApi? = null

  private var surfaceProducer: TextureRegistry.SurfaceProducer? = null
  private var renderThread: RenderThread? = null
  private var cameraSource: CameraSource? = null
  private var currentLayout: CompositeLayout? = null
  private var recordAudio = true
  private var hevc = false
  private var compositeW = 1280
  private var compositeH = 720
  private val mainHandler = Handler(Looper.getMainLooper())
  private val mainExecutor = Executor { mainHandler.post(it) }
  private var thermalListener: PowerManager.OnThermalStatusChangedListener? = null
  private var debugChannel: MethodChannel? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    context = binding.applicationContext
    textureRegistry = binding.textureRegistry
    flutterApi = DualCameraFlutterApi(binding.binaryMessenger)
    DualCameraHostApi.setUp(binding.binaryMessenger, this)
    debugChannel = MethodChannel(binding.binaryMessenger, DEBUG_CHANNEL).also {
      it.setMethodCallHandler(::handleDebug)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    releaseSession()
    DualCameraHostApi.setUp(binding.binaryMessenger, null)
    debugChannel?.setMethodCallHandler(null)
    debugChannel = null
    flutterApi = null
    textureRegistry = null
    context = null
  }

  // --- DualCameraHostApi ---

  override fun probeSupport(callback: (Result<CameraCapabilities>) -> Unit) {
    callback(Result.success(detectCapabilities()))
  }

  override fun initialize(
    config: RecordingConfig,
    callback: (Result<InitResult>) -> Unit,
  ) {
    val ctx = context
    val registry = textureRegistry
    if (ctx == null || registry == null) {
      callback(Result.failure(IllegalStateException("Plugin not attached")))
      return
    }
    try {
      val (w, h) = resolutionFor(config.resolution)
      compositeW = w
      compositeH = h
      recordAudio = config.recordAudio
      hevc = config.codec == VideoCodec.HEVC
      val layout = toCompositeLayout(config.layout, w, h)
      currentLayout = layout

      val producer = registry.createSurfaceProducer()
      producer.setSize(w, h)
      surfaceProducer = producer

      val render = RenderThread(w, h, layout)
      val source = CameraSource(ctx)
      render.listener = object : RenderThread.Listener {
        override fun onCameraSurfaces(back: Surface, front: Surface) {
          source.bind(
            back, front,
            onResolution = { isFront, sw, sh -> render.setSourceSize(isFront, sw, sh) },
            onRotation = { isFront, deg -> render.setSourceRotation(isFront, deg) },
            onError = { err ->
              flutterApi?.onError("camera_bind", err.message ?: "bind failed") {}
            },
          )
        }
      }
      render.statsListener = { fps, ms, dropped ->
        mainHandler.post {
          flutterApi?.onFrameStats(
            FrameStats(fps, ms, dropped, ThermalLevel.NOMINAL),
          ) {}
        }
      }
      render.setPreviewSurface(producer.surface, w, h)
      producer.setCallback(object : TextureRegistry.SurfaceProducer.Callback {
        override fun onSurfaceAvailable() {
          render.setPreviewSurface(producer.surface, w, h)
        }

        override fun onSurfaceDestroyed() {
          render.setPreviewSurface(null, w, h)
        }
      })
      renderThread = render
      cameraSource = source

      // Listener is now wired; hand over the camera surfaces to start the feed.
      render.startCameraFeed()

      registerThermalListener(ctx)
      val textureId = producer.id()
      flutterApi?.onReady(textureId) {}
      callback(Result.success(InitResult(textureId, detectCapabilities())))
    } catch (t: Throwable) {
      callback(Result.failure(t))
    }
  }

  /** Proactive thermal monitoring (ThermalGovernor, ARCHITECTURE.md §10). */
  private fun registerThermalListener(ctx: Context) {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return
    if (thermalListener != null) return
    val pm = ctx.getSystemService(Context.POWER_SERVICE) as? PowerManager ?: return
    val listener = PowerManager.OnThermalStatusChangedListener { status ->
      flutterApi?.onThermal(mapThermal(status)) {}
    }
    pm.addThermalStatusListener(mainExecutor, listener)
    thermalListener = listener
  }

  private fun unregisterThermalListener() {
    val ctx = context ?: return
    val pm = ctx.getSystemService(Context.POWER_SERVICE) as? PowerManager ?: return
    thermalListener?.let {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) pm.removeThermalStatusListener(it)
    }
    thermalListener = null
  }

  private fun mapThermal(status: Int): ThermalLevel = when {
    status >= PowerManager.THERMAL_STATUS_CRITICAL -> ThermalLevel.CRITICAL
    status == PowerManager.THERMAL_STATUS_SEVERE -> ThermalLevel.SERIOUS
    status == PowerManager.THERMAL_STATUS_MODERATE -> ThermalLevel.FAIR
    else -> ThermalLevel.NOMINAL
  }

  override fun startRecording(callback: (Result<Unit>) -> Unit) {
    val render = renderThread
    val ctx = context
    if (render == null || ctx == null) {
      callback(notYet("startRecording (not initialized)"))
      return
    }
    val path = File(ctx.cacheDir, "dcr_${System.nanoTime()}.mp4").absolutePath
    render.startRecording(path, recordAudio, hevc)
    flutterApi?.onRecordingStarted {}
    callback(Result.success(Unit))
  }

  override fun stopRecording(callback: (Result<String>) -> Unit) {
    val render = renderThread
    if (render == null) {
      callback(notYet("stopRecording (not initialized)"))
      return
    }
    render.stopRecordingResult { path ->
      mainHandler.post {
        flutterApi?.onRecordingStopped(path) {}
        callback(Result.success(path))
      }
    }
    return
  }

  override fun takePhoto(callback: (Result<String>) -> Unit) {
    val render = renderThread
    val ctx = context
    if (render == null || ctx == null) {
      callback(notYet("takePhoto (not initialized)"))
      return
    }
    val path = File(ctx.cacheDir, "dcr_${System.nanoTime()}.jpg").absolutePath
    render.takePhoto(
      path,
      done = { mainHandler.post { callback(Result.success(it)) } },
      fail = { mainHandler.post { callback(Result.failure(it)) } },
    )
  }

  override fun swapPrimary(callback: (Result<Unit>) -> Unit) {
    val layout = currentLayout ?: return callback(notYet("swapPrimary (not initialized)"))
    val swapped = layout.copy(primaryFront = !layout.primaryFront)
    currentLayout = swapped
    renderThread?.updateLayout(swapped)
    callback(Result.success(Unit))
  }

  override fun setLayout(layout: LayoutConfig, callback: (Result<Unit>) -> Unit) {
    if (surfaceProducer == null) return callback(notYet("setLayout (not initialized)"))
    val composite = toCompositeLayout(layout, compositeW, compositeH)
    currentLayout = composite
    renderThread?.updateLayout(composite)
    callback(Result.success(Unit))
  }

  override fun dispose(callback: (Result<Unit>) -> Unit) {
    releaseSession()
    callback(Result.success(Unit))
  }

  // --- debug tuning channel (example app only) ---

  /**
   * Live geometry tuning for the example harness — rotation correction, front
   * mirror, and source-aspect override — so the front/back orientation and any
   * stretch can be dialed in on a real device without rebuilding native code.
   */
  private fun handleDebug(call: MethodCall, result: MethodChannel.Result) {
    val render = renderThread
    if (render == null) {
      result.error("not_init", "initialize() first", null)
      return
    }
    when (call.method) {
      "setRotationOffset" -> {
        val front = call.argument<Boolean>("front") ?: true
        val offset = call.argument<Int>("offset") ?: 0
        render.setRotationOffset(front, offset)
        result.success(null)
      }
      "setAspectOverride" -> {
        val front = call.argument<Boolean>("front") ?: true
        val aspect = (call.argument<Double>("aspect") ?: 0.0).toFloat()
        render.setAspectOverride(front, aspect)
        result.success(null)
      }
      "setMirrorFront" -> {
        val on = call.argument<Boolean>("on") ?: true
        currentLayout?.let {
          val updated = it.copy(mirrorFront = on)
          currentLayout = updated
          render.updateLayout(updated)
        }
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  // --- internals ---

  private fun releaseSession() {
    unregisterThermalListener()
    cameraSource?.release()
    cameraSource = null
    renderThread?.release()
    renderThread = null
    surfaceProducer?.release()
    surfaceProducer = null
    currentLayout = null
  }

  private fun <T> notYet(op: String): Result<T> =
    Result.failure(UnsupportedOperationException("$op is not available"))

  // Portrait canvas (vertical video). Width < height; the cameras are rotated
  // upright into it by the compositor using each sensor's orientation.
  private fun resolutionFor(res: DualResolution): Pair<Int, Int> = when (res) {
    DualResolution.SD480 -> 480 to 854
    DualResolution.HD720 -> 720 to 1280
    DualResolution.HD1080 -> 1080 to 1920
  }

  private fun toCompositeLayout(layout: LayoutConfig, w: Int, h: Int): CompositeLayout {
    val primaryFront = layout.primary == CameraLens.FRONT
    val pip = layout.mode == DualLayoutMode.PICTURE_IN_PICTURE
    val splitVertical = layout.mode == DualLayoutMode.SPLIT_VERTICAL

    // Inset rect in normalized [0,1] surface coords (origin top-left).
    // insetScale is a fraction of the SHORTER side (height), so a circle inset
    // is a true square in pixels; a rect inset matches the canvas aspect.
    val insetH = layout.insetScale
    val insetW = if (layout.circleInset) layout.insetScale * h / w else layout.insetScale
    val mx = layout.margin / w
    val my = layout.margin / h
    val (l, t, r, b) = when (layout.insetCorner) {
      InsetCorner.TOP_LEFT -> arrayOf(mx, my, mx + insetW, my + insetH)
      InsetCorner.TOP_RIGHT -> arrayOf(1 - mx - insetW, my, 1 - mx, my + insetH)
      InsetCorner.BOTTOM_LEFT -> arrayOf(mx, 1 - my - insetH, mx + insetW, 1 - my)
      InsetCorner.BOTTOM_RIGHT ->
        arrayOf(1 - mx - insetW, 1 - my - insetH, 1 - mx, 1 - my)
    }
    // A circle is the rounded-rect SDF on a square rect with radius = half side.
    val insetSidePx = (insetH * h).toFloat()
    val cornerRadiusPx =
      if (layout.circleInset) insetSidePx / 2f else layout.cornerRadius.toFloat()
    return CompositeLayout(
      pictureInPicture = pip,
      primaryFront = primaryFront,
      insetLeft = l.toFloat(),
      insetTop = t.toFloat(),
      insetRight = r.toFloat(),
      insetBottom = b.toFloat(),
      cornerRadiusPx = cornerRadiusPx,
      mirrorFront = layout.mirrorFront,
      splitVertical = splitVertical,
    )
  }

  /**
   * Detect whether this device can run front + back concurrently. Mirrors
   * MASTER_PLAN.md §4.1.
   */
  private fun detectCapabilities(): CameraCapabilities {
    val ctx = context ?: return CameraCapabilities(
      isSupported = false,
      reason = UnsupportedReason.NO_CONCURRENT_CAMERA,
    )
    val manager = ctx.getSystemService(Context.CAMERA_SERVICE) as? CameraManager
      ?: return CameraCapabilities(
        isSupported = false,
        reason = UnsupportedReason.NO_CONCURRENT_CAMERA,
      )

    val supported = try {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
        manager.concurrentCameraIds.any { combo -> comboHasFrontAndBack(manager, combo) }
      } else {
        ctx.packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_CONCURRENT)
      }
    } catch (_: Throwable) {
      false
    }

    return if (supported) {
      CameraCapabilities(isSupported = true, maxWidth = 1280L, maxHeight = 720L)
    } else {
      CameraCapabilities(
        isSupported = false,
        reason = UnsupportedReason.NO_CONCURRENT_CAMERA,
      )
    }
  }

  private fun comboHasFrontAndBack(
    manager: CameraManager,
    combo: Set<String>,
  ): Boolean {
    var hasFront = false
    var hasBack = false
    for (id in combo) {
      val facing = try {
        manager.getCameraCharacteristics(id).get(CameraCharacteristics.LENS_FACING)
      } catch (_: Throwable) {
        null
      }
      when (facing) {
        CameraCharacteristics.LENS_FACING_FRONT -> hasFront = true
        CameraCharacteristics.LENS_FACING_BACK -> hasBack = true
      }
    }
    return hasFront && hasBack
  }

  private companion object {
    const val DEBUG_CHANNEL = "dual_cameras/debug"
  }
}
